core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+        verify(threadPool, times(1)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+
+        // create a thread and start the runnable
+        Thread runThread = new Thread() {
+            @Override
+            public void run() {
+                reschedulingRunnable.run();
+            }
+        };
+        runThread.start();
+
+        // wait for the runnable to be started and ensure the runnable hasn't used the threadpool again
+        startLatch.await();
+        verifyNoMoreInteractions(threadPool);
+
+        // un-pause the runnable and allow it to complete execution
+        pauseLatch.countDown();
+        runThread.join();
+
+        // validate schedule was called again
+        verify(threadPool, times(2)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+    }
+
+    public void testThatRunnableIsRescheduled() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(scaledRandomIntBetween(2, 16));
+        final Runnable countingRunnable = () -> {
+            if (rarely()) {
+                throw new ElasticsearchException("sometimes we throw before counting down");
+            }
+
+            latch.countDown();
+
+            if (randomBoolean()) {
+                throw new ElasticsearchException("this shouldn't cause the test to fail!");
+            }
+        };
+
+        Cancellable cancellable = threadPool.scheduleWithFixedDelay(countingRunnable, TimeValue.timeValueMillis(100L), Names.GENERIC);
+        assertNotNull(cancellable);
+
+        // wait for the number of successful count down operations
+        latch.await();
+
+        // cancel
+        cancellable.cancel();
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testCancellingRunnable() throws Exception {
+        final boolean shouldThrow = randomBoolean();
+        final AtomicInteger counter = new AtomicInteger(scaledRandomIntBetween(2, 16));
+        final CountDownLatch doneLatch = new CountDownLatch(1);
+        final AtomicReference<Cancellable> cancellableRef = new AtomicReference<>();
+        final AtomicBoolean runAfterDone = new AtomicBoolean(false);
+        final Runnable countingRunnable = () -> {
+
+            if (doneLatch.getCount() == 0) {
+                runAfterDone.set(true);
+                fail("this runnable was cancelled by itself so it should never run again!");
+            }
+
+            final Cancellable cancellable = cancellableRef.get();
+            if (cancellable == null) {
+                // wait for the cancellable to be present before we really start so we can accurately know we cancelled
+                return;
+            }
+
+            // rarely throw an exception before counting down
+            if (shouldThrow && rarely()) {
+                throw new RuntimeException("throw before count down");
+            }
+
+            final int count = counter.decrementAndGet();
+
+            if (shouldThrow && randomBoolean()) {
```

Comment:
what does this adds us in terms of testing coverage? isn't it just adding complexity?